### ISSUE TITLE
About template/accessibility-label-for

### ISSUE THREAD

**Issue by rafaelss95** [angular-eslint/angular-eslint#288]
I noticed some problems with `template/accessibility-label-for` and here are the suggestions (previously commented in #268):

1. It doesn't report labels without association, e.g.:

```html
<label for="nonExistentId">Label</label>
<!-- there's no element with `nonExistentId`, but the rule passes --> 
```

2. One could pass control component selectors through the `controlComponents` option, like this:

```json
"template-accessibility-label-for": [true, {
  "controlComponents": ["p-inputMask"],  
}]
```

```html
<app-label labelFor="username">Label<p-inputMask></p-inputMask></app-label>
```

...however one could have a component with an attribute that acts like an `id`, like [the `inputId` here, by example](https://www.primefaces.org/primeng/showcase/#/inputmask) and thus doesn't need to be wrapped by a label/custom label component. This case isn't covered by the rule, e.g.:

```html
<app-label labelFor="username">Label</app-label>
<p-inputMask formControlName="username" inputId="username" mask="(999) 999-9999? x99999"></p-inputMask>
```

3. The `labelAttributes` and `labelComponents` are stored in separated structures and it could be problematic, because a label attribute can cause conflicts depending on the label component you're using, e.g.:

```json
"template-accessibility-label-for": [true, {
  "labelAttributes": ["assoc", "elementId"],
  "labelComponents": ["app-label", "ngx-label"],  
}]
```

```html
<app-label assoc="id"></app-label> <!-- ok -->
<ngx-label assoc="id"></ngx-label > <!-- not ok, but the rule doesn't report -->
<app-label elementId="id"></app-label> <!-- not ok, but the rule doesn't report -->
```

<hr>

After thinking a little bit on how we could store the structures to meet these cases, I came up with this:

```js
"template-accessibility-label-for": [true, {
  "controlComponents": [
    { "inputs": ["id"], "selector": "input" }, // one of the defaults
    { "inputs": ["inputId"], "selector": "p-inputMask" },
    { "selector": "bs4-input" } // when it doesn't have `inputs`, it must be wrapped by a label/label component.
  ],
  "labelComponents": [
    { "inputs": ["for", "htmlFor"], "selector": "label" }, // default
    { "inputs": ["assoc"], "selector": "app-label" },
    { "inputs": ["elementId", "legacyId"], "selector": "ngx-label" },
    { "selector": "ng2-label" }, // when it doesn't have `inputs`, it must wrap a control component.
  ],
}]
```

To maintain the behavior in a schematic migration, the changes could be something like this:

```diff
"template-accessibility-label-for": ["error", {
- "controlComponents": ["p-inputMask", "bs4-input"],
- "labelAttributes": ["assoc", "elementId"],
- "labelComponents": ["app-label", "ngx-label"],
+ "controlComponents": [
+   { "selector": "p-inputMask" }, { "selector": "bs4-input" }
+ ],
+ "labelComponents": [
+   { "inputs": ["assoc", "elementId"], "selector": "app-label" },
+   { "inputs": ["assoc", "elementId"], "selector": "ngx-label" }
+ ],
}]
```

**Comment by JamesHenry** [comment]
@rafaelss95 I am not a user of this rule myself, but your approach looks comprehensive. Would you like to start on the implementation for v3?

I have a branch open for it already, including a migration schematic which you can add to

**Comment by JamesHenry** [comment]
Just noting for completeness that this was not a breaking change in the end, a new rule was added
